src/iarc7_sensors/vel_filter.py

```diff
@@ -0,0 +1,219 @@
+                return
+
+            if (math.isnan(new_vel_vector.x)
+             or math.isnan(new_vel_vector.y)
+             or math.isinf(new_vel_vector.x)
+             or math.isinf(new_vel_vector.y)):
+                rospy.logerr(
+                        'Velocity filter rejecting bad message: {}'.format(msg))
+                return
+
+            for i in xrange(len(self._queue) - 1, -1, -1):
+                if self._queue[i][1].header.stamp < msg.header.stamp:
+                    index = i+1
+                    break
+            else:
+                rospy.logerr(
+                        'Skipping message older than everything in the queue')
+                return
+
+            z_vel = float('NaN') if klass == 'optical' else new_vel_vector.z
+
+            self._queue.insert(index, (
+                    [float('NaN'), float('NaN'), z_vel],
+                    msg,
+                    klass)
+                )
+            self._reprocess(index)
+
+            # Make sure new timestamp is at least 0.1ms newer than the last one
+            new_stamp = max(self._last_published_stamp + rospy.Duration(0, 100),
+                            self._queue[-1][1].header.stamp)
+
+            last_kalman_i = self._get_last_index('kalman')
+            if last_kalman_i == -1 or not self._position_passthrough:
+                self._publish_odom(0.0,
+                                   0.0,
+                                   0.0,
+                                   *self._queue[-1][0],
+                                   time=new_stamp)
+            else:
+                last_kalman_msg = self._queue[last_kalman_i][1]
+                self._publish_odom(last_kalman_msg.pose.pose.position.x,
+                                   last_kalman_msg.pose.pose.position.y,
+                                   last_kalman_msg.pose.pose.position.z,
+                                   *self._queue[-1][0],
+                                   time=new_stamp)
+
+            self._last_published_stamp = new_stamp
+
+            while len(self._queue) > self._message_queue_length:
+                self._queue.pop(0)
+
+    def _reprocess(self, index):
+        for i in xrange(index, len(self._queue)):
+            msg = self._queue[i][1]
+            klass = self._queue[i][2]
+
+            # Get index of last differential measurement in the queue
+            last_differential_i = self._get_last_index('kalman',
+                                                       msg.header.stamp)
+            assert last_differential_i < i and last_differential_i >= -1
+
+            if klass == 'kalman':
+                if last_differential_i == -1:
+                    # No other differential measurements in the queue, so
+                    # assume our velocity hasn't changed since the previous
+                    dx = 0.0
+                    dy = 0.0
+                else:
+                    last_diff_vector = \
```

Comment:
What is this operation doing? I've never seen that before. @amiller27

Reply:
Backslash is line continuation in Python, it just means the next line isn't a new statement. You only need them if you aren't currently inside parens or brackets.

Reply:
Oh okay makes sense.